[Time Conductor] Support non-UTC domains

VWoeltjen opened the pull request:
Adds support for non-UTC (and non-time-like) domains to the time conductor, permitting handling of Mars time and other distinct forms.

Summary of changes:
* Add a `TimeSystem` interface, which provides generic details about how to interpret values for different time systems.
* Add a `TimeService` interface, for composite services which provide the above
* Implement a `TimeService` aggregator and a provider for UTC 
* Add a `system` property to telemetry domain metadata
  * Importantly, this is distinct from domains themselves: 
  * Assume UTC (in millis since 1970) if no time system is specified.
* Add JSDoc for telemetry metadata to reflect the above

Still a work in progress; remaining to do:
- [x] Add tests for classes in ~~platform/time~~ `platform/commonUI/formatting` and other new scripts.
- [x] Add JSDoc for new interfaces/classes
- [x] Utilize time service wherever timestamps from telemetry are formatted for display
- [x] Update tests for modified components
- [x] Separate date-time text-input/pickers into an `mct-control` (per feedback from #202)
- [x] Remove `now`, increment methods from `TimeSystem`; rename to `DateFormat` or `Format`
- [x] Expose different time systems via an extension category instead of composite services
- [x] ~~Add a `domainService` which includes both domain options and their formats.~~ Don't see this as necessary at this point, `TIME_CONDUCTOR_DOMAINS` plus telemetry metadata remains sufficient.

Opening pull request early for feedback (esp. from @larkin); in particular:

* Not feeling great about interface depth (adding both `TimeService` and `TimeSystem`), but sticking all these methods on `TimeService` (along with a parameter identifying the time system) made both using it and implementing it awkward.
* Naming: Is there a better name than `TimeSystem`, `TimeService`? 
  * Some single-word alternatives to `TimeSystem` include:
    * `Chronology`
    * `Temporality`
    * `Continuity`
    * `Continuum` 
    * `Calendar`
  * One of the challenges with naming here is that these are almost-but-not-always time-like; counters and the like will need to be dealt with here. `Continuum` seems most accurate in that regard but is at best confusing and at worst confounding.
* Should `TimeSystem` instances just be a new extension category?

larkin commented on example/generator/src/SinewaveTimeSystem.js:
```diff
@@ -0,0 +1,55 @@
+/*global define*/
+
+define([
+    './SinewaveTelemetrySeries'
```
Unused.

VWoeltjen replied:
Used in the `now` method, but based on our recent conversations about LST in #182 I'm inclined to replace this with a more time-like (but still non-date-formatted) time system

larkin replied:
Ah, I totally missed that.  I guess this is a DRY method of handling it, but a little confusing when trying to wrap one's mind around "what does a time system mean?"

larkin commented on platform/time/bundle.json:
```diff
@@ -0,0 +1,19 @@
+                "implementation": "TimeAggregator.js"
+            },
+            {
+                "provides": "timeService",
```
Seems like the composite service pattern is overkill when we're really looking for a type registry, and I'm concerned it is more confusing when we implement aggregators that expose a different interface than the interface of the individual providers.  (I have been known to do this... say in the search aggregator).

What do you think about using a new extension category for time formats instead of composite services?

larkin commented on platform/time/src/UTCTimeProvider.js:
```diff
@@ -0,0 +1,44 @@
+        this.utcTimeSystem = new UTCTimeSystem(now);
+    }
+
+    UTCTimeProvider.prototype.systems = function () {
```
This strikes me as a weird way to register the name of the time system that this provides times for.

larkin commented:
@VWoeltjen this does indeed have a significant number of interfaces for what it provides.  I think an extension category would be a better way to expose time formats, and if the array return value that one gets from depending on a extension category is not ideal, a time service could depend on the extension category and generate a object for key-based lookup of time systems.

larkin commented:
@VWoeltjen as for your naming question, if a `TimeSystem` has the following methods:

* `parse`
* `format`
* `validate`
* `now`

then, if we factor out the `now` method -- by assuming that `now` is the same value for all time systems-- then what we would be left with is:

* `parse`
* `format`
* `validate`

Which I would happily call a `DateFormat`.

After looking at it this way, I'd rather keep the concept of `now` outside of the date format.  Perhaps we have a time system object which includes both a date format and some method for determining `now`, but I don't see the need for that now (tee hee).

VWoeltjen commented:
Agree, doing this as an extension category makes sense. Have been avoiding adding more of those based on feedback that there were too many, but really, this is the sort of situation they were put there for. (Making the "too many" less daunting or easier to deal with is really a problem for #69.)

I agree that this would be simpler as a `DateFormat`, but where we get into trouble is interpreting LST values. These are relative to mission start, but taking this approach would require us to convert those to UTC times basically any time they're not being formatted for display. I suppose it is not much harder to do this than it is to come up with a decent value for `now`, though.

larkin commented:
One final comment:  I think it would help if we separated out the "business logic" of the time conductor from the generic date objects used.  

That is, force those who would like to use the time conductor to implement a "domain service" which is used by the time conductor.  Such a service would provide a list of domains available for the time conductor, and for each domain, methods for determining the default interval to use.  Domains would have a "format" which would map to a DateFormat, but would impose nothing upon date formats.

VWoeltjen commented:
Ready for review. Change summary is described above; only thing to add is that I made some changes to plot to partially address #218 (known issue, moved from closed-source issue tracker); that is, made plot aware enough of domains to utilize formats by fixing some of the ordering problems that contributed to #218. Domain options are now refreshed regularly (perhaps too often, but don't see an impact when profiling) to ensure appropriate formats are used in plot when domain selected from time conductor changes.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Expect to pass code review? Y

larkin commented on platform/commonUI/formats/test/UTCTimeFormatSpec.js:
```diff
@@ -0,0 +1,56 @@
+            });
+
+            it("validates time inputs", function () {
+                expect(format.validate("1977-05-25 11:21:22")).toBeTruthy();
```
Minor, but when a function returns a Boolean, I would prefer specs to be precise, e.g `.toBe(true)` or `.toBe(false)`; your thoughts?

VWoeltjen replied:
Makes sense, think that specificity is useful here. There are a few cases (e.g. flags in scope, like `DateTimeFieldController` in this PR) where I think that would be over-specifying, but for documented return types, definitely appropriate.

larkin commented on platform/commonUI/formats/src/FormatProvider.js:
```diff
@@ -0,0 +1,108 @@
+
+        function addToMap(Format) {
+            var key = Format.key;
+            if (key && !formatMap[key]) {
```
Question-- does this prevent overriding of a format by different bundles, or would standard bundle priorities still determine which format is loaded for a given key?

VWoeltjen replied:
Yep, this obeys the normal priority convention for extensions (so you could override `utc` by setting priority to "preferred", for example)

larkin replied:
:+1:

larkin commented on platform/commonUI/formats/src/UTCTimeFormat.js:
```diff
@@ -0,0 +1,64 @@
+    var DATE_FORMAT = "YYYY-MM-DD HH:mm:ss",
+        DATE_FORMATS = [
+            DATE_FORMAT,
+            "YYYY-MM-DD HH:mm:ss",
```
Is this a duplicate of DATE_FORMAT?

VWoeltjen replied:
Yep - pushed a new commit to remove the duplicate

larkin commented on platform/commonUI/general/src/controllers/TimeRangeController.js:
```diff
@@ -243,36 +211,12 @@ define(
-                }
-            }
+            function updateFormat(key) {
+                formatter = formatService.getFormat(key) ||
```
Hrm, wondering if it's wise to fallback to utc when a requested format isn't available-- would it be better to throw an error-- perhaps at the level of the format service?  

Also, at the level of the time conductor, it should always be getting the format set by the selected domain, correct?  So it shouldn't need to have a separate "default format".

VWoeltjen replied:
I see two cases here:

* Format has not been specified. I'd like to handle this case by falling back to `utc` as a default; this is what will apply to the majority of telemetry data. Thinking of the developer who is just trying to get some "hello, world" telemetry going, don't want them to get stifled by one missing property. Considering this more, though, I think it would make more sense to supply these defaults from `TelemetryCapability` so that other components can just assume the property will be there.
* Format has been specified but is not recognized. Think you're right that we shouldn't _just_ fall back to UTC here; I see two options:
  * Log a warning and fall back to the default.
  * Throw an error and just fail.

I like falling back to a warning because it's more fault-tolerant, but on consideration I think there is a correctness concern there - you'd be showing data as dates when it's really meant to be displayed as some other thing entirely.

Think this is true for `DateTimeFieldController` as well.

larkin replied:
For the various mct-control definitions, having defaults for when a domain is not provided makes sense; as that's just configuring an input field, and I can see people wanting to control that, or optionally not care about it.

As you say, allowing invalid domains to be used is a correctness concern and should be prevented-- if a user thinks they are entering LST but they're really entering UTC, they would see significantly incorrect data.

larkin commented on platform/commonUI/general/src/controllers/DateTimeFieldController.js:
```diff
@@ -0,0 +1,76 @@
+            var formatter = formatService.getFormat(DEFAULT_FORMAT);
+
+            function setFormat(format) {
+                formatter = formatService.getFormat(format) ||
```
At this level, I think having a default format makes sense if `structure.format` is undefined, but having a fallback when a specified format is unavailable seems like it should throw an error.

larkin commented on platform/features/conductor/bundle.json:
```diff
@@ -36,9 +36,9 @@
                 "value": [
-                    { "key": "time", "name": "Time" },
-                    { "key": "yesterday", "name": "Yesterday" }
+                    { "key": "time", "name": "Time" }
```
Thinking a better default here would be 'UTC' and not 'Time'.

larkin commented on platform/features/conductor/src/ConductorRepresenter.js:
```diff
@@ -74,11 +77,12 @@ define(
                     start: conductor.displayStart(),
                     end: conductor.displayEnd(),
-                    domain: conductor.domain()
+                    domain: domain && domain.key
```
Is it ever valid that conductor.domain() would not return an object?

larkin commented on platform/features/conductor/src/ConductorRepresenter.js:
```diff
@@ -97,12 +101,10 @@ define(
-                ));
+                var newDomain = conductor.domain(value);
+                conductorScope.parameters.format =
+                    newDomain && newDomain.format;
```
Similar to above-- is it valid for conductor.domain() not to return a domain object?  (isn't this part of the definition of `conductor.domain`)?

VWoeltjen replied:
I had a moment of confusion between "domain object" and "domain object" there :grinning: 

As implemented, no. I was thinking ahead to the case where `TimeConductor` might be populated by metadata from capabilities, but even in that case, I'm thinking now that it should be `TelemetryCapability`'s responsibility to fill in defaults so that user code can just assume complete metadata is available. (I might try to stick those changes into a follow up issue and address only the subset that's needed to simplify this change set.)

larkin replied:
Still falls back on undefined and would cause unexpected behavior here.

larkin commented on platform/features/conductor/src/ConductorRepresenter.js:
```diff
@@ -130,7 +132,8 @@ define(
             conductorScope.ngModel.options =
                 conductor.domainOptions().map(makeOption);
-            conductorScope.ngModel.domain = conductor.domain();
+            conductorScope.ngModel.domain = (conductor.domain() || {}).key;
```
Again-- is there a case where this won't return a valid object?

larkin commented on platform/features/conductor/src/ConductorTelemetryDecorator.js:
```diff
@@ -51,7 +51,7 @@ define(
                 request.start = start;
                 request.end = end;
-                request.domain = domain;
+                request.domain = domain && domain.key;
```
It seems to me that at this level, we're hiding invalid input errors that should actually cause the caller to receive an error; calling requestTelemetry without an array is invalid, and providing an array that contains items that are not objects is also invalid.  Likewise, the conductor should always be providing a real domain.

larkin commented:
@VWoeltjen done with that pass-- my primary concern is there are many checks for invalid values that allow program flow to silently proceed where invalid values should never be provided; see this as troublesome for debugging issues later.  Should be easy to remove these and then otherwise, looks good to ship.

Can you address and assign back to me for R&I?

VWoeltjen commented:
Bullets to address from code review:

- [x] Increase test specificity (`toBe(true)` instead of `toBeTruthy()`)
- [x] Remove redundant date format
- [ ] Remove checks for undefined `domain` returned from `TimeConductor`
- [ ] Change default domain name to "UTC" (from "Time")
- [ ] Default to `utc` only when unspecified
- [ ] Throw errors when a specified format is not found
- [ ] File follow-up issue to apply defaults to metadata from `TelemetryCapability`

VWoeltjen commented:
Addressed feedback from code review as described in the previous comment. Additional notes:

* Instead of throwing errors for unknown formats in `telemetryFormatter`, just coerce the value to a string (to display the raw numeric value)
  * This is different from components of the time conductor (which just throw errors.) Rationale is that `telemetryFormatter` has actual data to display, so displaying it in its raw form should not be misleading (although it may be less useful)
* Observed cases where format changes were not updating displayed values data-time fields; added an update to these values whenever format changes.
* Added some JSDoc for parameters that was missing

Reassigning back over to @larkin to review updates

larkin commented:
@VWoeltjen Why not throw errors in the format service for any non-existent format?  Allowing `formatService.getFormat` to have two possible return values requires other components to handle that potential behavior.  With a simpler contract, they don't have to perform such checks, and the complexity of the software is decreased.

If we decided that such error handling was necessary in the future, we could implement it, but I'd rather back it up with a real use case; I don't see a case where a developer would add a time format without the supporting code.

VWoeltjen commented:
@larkin - semantically, returning "undefined" seemed more natural / less unexpected for a lookup-style service. I'm not attached.

I think there's a bit of funniness in here insofar as the errors we're talking about are very configuration-driven in practice. Thinking it might be better to fold this into the `telemetry` capability in the long run, and make that responsible for providing everything you need to work with telemetry for the domain object. This would allow you to throw exceptions (or log warnings) on misconfiguration much sooner. Added a note in #69 about this.

In any event, pushed updates which throw errors from getFormat, and to remove the missed check for an undefined return value from `TimeConductor#domain`

larkin commented:
# Review Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
5. Project-specific information isolated to appropriate branches? Y

# Integration Checklist

1. Automated build passing after merge? Y
2. No merge conflicts (or resolution trivial/obvious)? Y
3. All master branches up-to-date after merge? Y
4. Project-specific information isolated to appropriate branches? Y
